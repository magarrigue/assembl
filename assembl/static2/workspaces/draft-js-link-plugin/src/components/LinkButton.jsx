// @flow
import * as React from 'react';
import EditorUtils from 'draft-js-plugins-utils';
import ReactDOM from 'react-dom';
import { I18n } from 'react-redux-i18n';

import AddLinkForm from './AddLinkForm';
import PluginModal from './PluginModal';
import type { DraftJSPluginStore, Theme } from '../index';

type Props = {
  modal: ?{ current: null | React.ElementRef<any> },
  store: DraftJSPluginStore,
  theme: Theme
};

type State = {
  showModal: boolean
};

export default class LinkButton extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      showModal: false
    };
  }

  onMouseDown = (event: SyntheticMouseEvent<HTMLDivElement>) => {
    event.preventDefault();
  };

  addLink = (values: { text: string, url: string }) => {
    if (this.props.store) {
      const { getEditorState, setEditorState } = this.props.store;
      if (getEditorState && setEditorState) {
        const { url } = values;
        // TODO: handle openInNewTab in createLinkAtSelection
        setEditorState(EditorUtils.createLinkAtSelection(getEditorState(), url));
        this.setState({ showModal: false });
      }
    }
  };

  openModal = (e: SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    e.stopPropagation();
    this.setState({ showModal: true });
  };

  closeModal = (e: SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    e.stopPropagation();
    this.setState({ showModal: false });
  };

  getSelectedText = () => {
    const { store } = this.props;
    if (!store.getEditorState) {
      return '';
    }

    const editorState = store.getEditorState();
    const selectionState = editorState.getSelection();
    const anchorKey = selectionState.getAnchorKey();
    const currentContent = editorState.getCurrentContent();
    const currentContentBlock = currentContent.getBlockForKey(anchorKey);
    const start = selectionState.getStartOffset();
    const end = selectionState.getEndOffset();
    const selectedText = currentContentBlock.getText().slice(start, end);
    return selectedText;
  };

  render() {
    const { modal, theme } = this.props;
    const modalContainer = modal && modal.current;
    return (
      <React.Fragment>
        {modalContainer &&
          this.state.showModal &&
          ReactDOM.createPortal(
            <PluginModal
              close={this.closeModal}
              closeLabel={I18n.t('common.editor.linkPlugin.addLinkForm.close')}
              title={I18n.t('common.editor.linkPlugin.addLinkForm.title')}
            >
              <AddLinkForm defaultText={this.getSelectedText()} onSubmit={this.addLink} />
            </PluginModal>,
            modalContainer
          )}
        <div className={theme.buttonWrapper} onMouseDown={this.onMouseDown}>
          <button className={theme.button} onClick={this.openModal} type="button">
            <span className="assembl-icon-text-link" />
          </button>
        </div>
      </React.Fragment>
    );
  }
}