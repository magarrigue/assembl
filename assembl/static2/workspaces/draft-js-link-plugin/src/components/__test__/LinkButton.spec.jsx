import React from 'react';
import renderer from 'react-test-renderer';

import LinkButton from '../LinkButton';

describe('LinkButton component', () => {
  it('should render a link button', () => {
    const getEditorStateSpy = jest.fn();
    const setEditorStateSpy = jest.fn();
    const props = {
      modal: {
        current: null
      },
      store: {
        getEditorState: getEditorStateSpy,
        setEditorState: setEditorStateSpy
      },
      theme: {
        button: 'btn',
        buttonWrapper: 'btn-group'
      }
    };
    const component = renderer.create(<LinkButton {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // TODO: test that modal is rendered in portal
  it('should render a link button with an opened modal');
});