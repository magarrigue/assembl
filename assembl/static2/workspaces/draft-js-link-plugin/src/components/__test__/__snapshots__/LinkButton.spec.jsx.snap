// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkButton component should render a link button 1`] = `
<div
  className="btn-group"
  onMouseDown={[Function]}
>
  <button
    className="btn"
    onClick={[Function]}
    type="button"
  >
    <span
      className="assembl-icon-text-link"
    />
  </button>
</div>
`;
